File: examples/governance/compliance_domain/lib/compliance_domain/exemption/commands/revoke_exemption.rb
module ComplianceDomain
  class Exemption
    module Commands
      class RevokeExemption
        include Hecks::Command
        emits "RevokedExemption"

        attr_reader :exemption_id

        def initialize(exemption_id: nil)
          @exemption_id = exemption_id
        end

        def call
          existing = repository.find(exemption_id)
          if existing
            unless existing.status == "active"
              raise ComplianceDomain::Error, "Cannot RevokeExemption: status must be 'active', got '#{existing.status}'"
            end
            Exemption.new(
              id: existing.id,
              model_id: existing.model_id,
              policy_id: existing.policy_id,
              requirement: existing.requirement,
              reason: existing.reason,
              approved_by_id: existing.approved_by_id,
              approved_at: existing.approved_at,
              expires_at: existing.expires_at,
              scope: existing.scope,
              status: "revoked"
            )
          else
            raise ComplianceDomain::Error, "Exemption not found: #{exemption_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/revoke_exemption.go
package domain

import (
	"time"
	"fmt"
)

type RevokeExemption struct {
	ExemptionId string `json:"exemption_id"`
}

func (c RevokeExemption) CommandName() string { return "RevokeExemption" }

func (c RevokeExemption) Execute(repo ExemptionRepository) (*Exemption, *RevokedExemption, error) {
	existing, err := repo.Find(c.ExemptionId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Exemption not found: %s", c.ExemptionId)
	}
	if existing.Status != "active" {
		return nil, nil, fmt.Errorf("cannot RevokeExemption: Exemption is in %s state", existing.Status)
	}
	existing.Status = "revoked"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RevokedExemption{
		AggregateID: existing.ID,
		ExemptionId: c.ExemptionId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
